ndb/src/google/cloud/ndb/model.py

```diff
@@ -2947,28 +2967,405 @@ def __init__(self, *args, **kwargs):
+    class name (exclusive of the module name or any other parent scope). To
+    override the kind, define :meth:`_get_kind`, as follows:
+
+    .. code-block:: python
+
+        class MyModel(ndb.Model):
+            @classmethod
+            def _get_kind(cls):
+                return "AnotherKind"
+
+    A newly constructed entity will not be persisted to Cloud Datastore without
+    an explicit call to :meth:`put`.
+
+    User-defined properties can be passed to the constructor via keyword
+    arguments:
+
+    .. doctest:: model-keywords
+
+        >>> class MyModel(ndb.Model):
+        ...     value = ndb.FloatProperty()
+        ...     description = ndb.StringProperty()
+        ...
+        >>> MyModel(value=7.34e22, description="Mass of the moon")
+        MyModel(description='Mass of the moon', value=7.34e+22)
+
+    In addition to user-defined properties, there are six accepted keyword
+    arguments:
+
+    * ``key``
+    * ``id``
+    * ``app``
+    * ``namespace``
+    * ``parent``
+    * ``projection``
+
+    Of these, ``key`` is a public attribute on :class:`Model` instances:
+
+    .. testsetup:: model-key
+
+        from google.cloud import ndb
+
+
+        class MyModel(ndb.Model):
+            value = ndb.FloatProperty()
+            description = ndb.StringProperty()
+
+    .. doctest:: model-key
+
+        >>> entity1 = MyModel(id=11)
+        >>> entity1.key
+        Key('MyModel', 11)
+        >>> entity2 = MyModel(parent=entity1.key)
+        >>> entity2.key
+        Key('MyModel', 11, 'MyModel', None)
+        >>> entity3 = MyModel(key=ndb.Key(MyModel, "e-three"))
+        >>> entity3.key
+        Key('MyModel', 'e-three')
+
+    However, a user-defined property can be defined on the model with the
+    same name as one of those keyword arguments. In this case, the user-defined
+    property "wins":
+
+    .. doctest:: model-keyword-id-collision
+
+        >>> class IDCollide(ndb.Model):
+        ...     id = ndb.FloatProperty()
+        ...
+        >>> entity = IDCollide(id=17)
+        >>> entity
+        IDCollide(id=17.0)
+        >>> entity.key is None
+        True
+
+    In such cases of argument "collision", an underscore can be used as a
+    keyword argument prefix:
+
+    .. doctest:: model-keyword-id-collision
+
+        >>> entity = IDCollide(id=17, _id=2009)
+        >>> entity
+        IDCollide(key=Key('IDCollide', 2009), id=17.0)
+
+    For the **very** special case of a property named ``key``, the ``key``
+    attribute will no longer be the entity's key but instead will be the
+    property value. Instead, the entity's key is accessible via ``_key``:
+
+    .. doctest:: model-keyword-key-collision
+
+        >>> class KeyCollide(ndb.Model):
+        ...     key = ndb.StringProperty()
+        ...
+        >>> entity1 = KeyCollide(key="Take fork in road", id=987)
+        >>> entity1
+        KeyCollide(_key=Key('KeyCollide', 987), key='Take fork in road')
```

Comment:
@chrisrossi In the existing implementation the behavior is slightly different:

```python
from google.appengine.ext import ndb


class KeyCollide(ndb.Model):
    key = ndb.StringProperty()


entity1 = KeyCollide(key="Take fork in road", id=987)
print(entity1)
entity2 = KeyCollide(key="Go slow", _key=ndb.Key(KeyCollide, 1))
print(entity2)
entity3 = KeyCollide(key="Zap")
print(entity3)
entity4 = KeyCollide(id="Baz")
print(entity4)
entity5 = KeyCollide()
print(entity5)
```

yields

```python
KeyCollide(key=Key('KeyCollide', 987), key='Take fork in road')
KeyCollide(key=Key('KeyCollide', 1), key='Go slow')
KeyCollide(key='Zap')
KeyCollide(key=Key('KeyCollide', 'Baz'))
KeyCollide()
```

Reply:
Your fix is fine. repr() should theoretically return text that if it were python would instantiate an equivalent object. The old implementation failed that test.

Reply:
:+1: